### PR TITLE
Queen bulwark

### PR DESCRIPTION
**Pull request by TiviPlus**
## About The Pull Request
Places a channeled AOE around the queen that while maintained grants all xenos in it +25% armor(based on their default armor)

VFX for xenos maybe should be added but idk we'll see

## Why It's Good For The Game
Now that I've cleaned up some queen real estate it can be used to add stuff so queen has more to do than hit screech

## Changelog
:cl:
add: Added Bulwark ability to queen
/:cl:

### CODE REVIEW

**Comment by Arturlang** [comment]
Does this mean the queen has to stay still to apply a 25 armor buff to xenos? Does this include the queen themselves? If forces the queen to stay still, then it's not going to be very useful at all, forcing you to become a target for a slight buff for xenos.

**Comment by Losenis** [comment]
>Crusher showing up after queen ran in and popped screech + bulwark with ward pheros
arnold-schwarzenegger-meme

**Comment by maikilangiolo** [comment]
> not going to be very useful at all
> a slight buff for xenos.

It's a huge buff. armor isn't linear, a 5 armor increase is already a lot.

**Comment by Losenis** [comment]
> then it's not going to be very useful at all
it's going to be extremely useful when queen runs in and screeches for the rest of the hive to dogpile in, since as it currently stands right now the queen can only just spit and slash whereas now it'd probably give the entire hive way more survivability to get a shitton of damage done before each is forced to retreat, and this isn't even taking into account ward pheros that'll stack.

An ancient crusher has 75 base armor and with this it'd go up to 93.75. It's a 25% increase in armor but it reduces the damage you get down to a fourth, equivalent from having a larva go from 0 armor to 75 armor. A ravager will go from 55 armor to 68.75, which translates to effectively getting 30% less damage if my math is right. Of course all of it is based on damage without pen but still, queen goes in, screeches, and suddenly everything around it gets a 25% armor buff? The massacre will be absolutely hilarious.

**Comment by Araulius** [comment]
more abilities that encourage team play is always welcome

**Comment by bowlofcereal** [comment]
i think giving the queen more versatility to feel more like the leader then just spitslash and scream at people is a good idea but that 5 armor increase is wicked goddamn

**Comment by iLikeMemes** [comment]
Queen already has a lot of abilities and i feel like this is overkill. Although an armor increasing ability does sound nice, this doesnt feel like the right way to implement it imo.

**Comment by Lumipharon** [comment]
Increasing armour scales heavily for high armour castes. Stacked king getting this buff is effectively reducing incoming damage by 50%, for normal guns.

**Comment by Strayed-Raven** [comment]
Re-reading this over and over just keeps giving me the impression its a change from a raffle ticket game.
What a path to take.

**Comment by ilikesaying** [comment]
Does this affect Defender’s crest and fortify too?

**Comment by Joe13413** [comment]
i've always wanted more abilities for queen so that it feels more fun, but this just seems WAY overkill. the problem IMO is the fact that queen doesn't really have super fun combat OR support loops (combat is boring, support skills dont feel super impactful) so i'd probably stick to improving one of those

**Review by Kuro020 — APPROVED** [review]

**Comment by ilikesaying** [comment]
does it stack with or on top of defender's fortify

File: code/modules/mob/living/carbon/xenomorph/castes/queen/abilities_queen.dm [hunk 1]
```diff
@@ -439,3 +439,86 @@
...
+		UnregisterSignal(target, list(COMSIG_ATOM_EXITED, COMSIG_ATOM_ENTERED))
+		REMOVE_TRAIT(target, TRAIT_BULWARKED_TURF, XENO_TRAIT)
+		for(var/mob/living/carbon/xenomorph/xeno AS in armor_mod_keys)
+			remove_buff(null, xeno)
+	affected_turfs = null
+
+///adds buff to xenos
+/datum/action/xeno_action/bulwark/proc/apply_buff(datum/source, mob/living/carbon/xenomorph/xeno, direction)
+	SIGNAL_HANDLER
+	if(!isxeno(xeno) || armor_mod_keys[xeno] || !owner.issamexenohive(xeno))
```
Comment (QualityVan) [inline code/modules/mob/living/carbon/xenomorph/castes/queen/abilities_queen.dm:491]: Should the first part be an istype check? Does it matter?

**Comment by breadhunt** [comment]
Imo queen not having a shitton of abilities is fine, they should probably be leading. Not to mention they already have spitslash, stickyspit, weeds/walls, screech, acid wells, plasma buffs, heals... 
Also if this forces queen to channel and stay still while other xenos go rip and rear won't this *decrease* the fun value of queenie?

File: code/modules/mob/living/carbon/xenomorph/castes/queen/abilities_queen.dm [hunk 2]
```diff
@@ -439,3 +439,86 @@
...
+#define BULWARK_ARMOR_MULTIPLIER 1.25
+
+/datum/action/xeno_action/bulwark
+	name = "Royal Bulwark"
+	action_icon_state = "bulwark"
+	desc = "Creates a field of defensive energy, filling chinks in the armor of nearby sisters, making them more resilient."
+	plasma_cost = 100
+	cooldown_timer = 20 SECONDS
+	keybinding_signals = list(
+		KEYBINDING_NORMAL = COMSIG_XENOABILITY_QUEEN_HIVE_MESSAGE,
```
[the remark this answers is not in the document]
Comment (TiviPlus) [reply code/modules/mob/living/carbon/xenomorph/castes/queen/abilities_queen.dm:455]: fixed

File: code/modules/mob/living/carbon/xenomorph/castes/queen/abilities_queen.dm [hunk 3]
```diff
@@ -439,3 +439,86 @@
...
+	affected_turfs = null
+
+///adds buff to xenos
+/datum/action/xeno_action/bulwark/proc/apply_buff(datum/source, mob/living/carbon/xenomorph/xeno, direction)
+	SIGNAL_HANDLER
+	if(!isxeno(xeno) || armor_mod_keys[xeno] || !owner.issamexenohive(xeno))
+		return
+	var/datum/armor/basearmor = getArmor(arglist(xeno.xeno_caste.soft_armor))
+	var/datum/armor/maxarmor = basearmor.scaleAllRatings(BULWARK_ARMOR_MULTIPLIER)
+	var/datum/armor/armordiff = maxarmor.detachArmor(basearmor)
```
[the remark this answers is not in the document]
Comment (TiviPlus) [reply code/modules/mob/living/carbon/xenomorph/castes/queen/abilities_queen.dm:495]: done

File: code/modules/mob/living/carbon/xenomorph/castes/queen/abilities_queen.dm (hunk 1 above)
Comment (TiviPlus) [reply code/modules/mob/living/carbon/xenomorph/castes/queen/abilities_queen.dm:491]: Yea, entered signal handler

**Review by Kuro020 — APPROVED** [review]

**Review by Lumipharon — APPROVED** [review]
Code's good, still think this is questionable balance wise since you can have 100+ armour xenos but eh.

File: code/modules/mob/living/carbon/xenomorph/castes/queen/abilities_queen.dm [hunk 4]
```diff
@@ -448,3 +448,85 @@
...
+	var/list/turf/affected_turfs = RANGE_TURFS(BULWARK_RADIUS, owner)
+	add_cooldown()
+
+	for(var/turf/target AS in affected_turfs)
+		//yes I realize this adds and removes it every move but its simple
+		//also we use this and not aura because we want speedy updates on entering
+		RegisterSignal(target, COMSIG_ATOM_EXITED, PROC_REF(remove_buff))
+		RegisterSignal(target, COMSIG_ATOM_ENTERED, PROC_REF(apply_buff))
+		ADD_TRAIT(target, TRAIT_BULWARKED_TURF, XENO_TRAIT)
+		for(var/mob/living/carbon/xenomorph/xeno in target)
```
Comment (Lumipharon) [inline code/modules/mob/living/carbon/xenomorph/castes/queen/abilities_queen.dm:479]: is it easier to just range check the hive list? That would also mean you're not picking up other hives, in the blue moon that's relevant.

File: code/modules/mob/living/carbon/xenomorph/castes/queen/abilities_queen.dm [hunk 5]
```diff
@@ -448,3 +448,85 @@
...
+		succeed_activate()
+
+	aoe_particles.particles.spawning = 0
+	QDEL_IN(aoe_particles, 4 SECONDS)
+
+	for(var/turf/target AS in affected_turfs)
+		UnregisterSignal(target, list(COMSIG_ATOM_EXITED, COMSIG_ATOM_ENTERED))
+		REMOVE_TRAIT(target, TRAIT_BULWARKED_TURF, XENO_TRAIT)
+		for(var/mob/living/carbon/xenomorph/xeno AS in armor_mod_keys)
+			remove_buff(null, xeno)
```
Comment (Lumipharon) [inline code/modules/mob/living/carbon/xenomorph/castes/queen/abilities_queen.dm:494]: extra indent

**Review by Lumipharon — CHANGES_REQUESTED** [review]

**Comment by Lumipharon** [comment]
image

Concern.jpg

**Comment by novaepee** [comment]
After the armor nerf, increasing the range for bulwark would be great. Ivanmixo expressed that the range leaves much to be desired.

**Comment by Texicos** [comment]
Ability is almost useless with how prevalent sunder is with lasguns being the new "meta"

**Comment by maikilangiolo** [comment]
Just let this die.

**Comment by Lumipharon** [comment]
it was tm'd for less than a day because it conflicts with shrike.

**Pull request merged**